### PR TITLE
Fix Consistency Models UNet2DMidBlock2D Attention GroupNorm Bug

### LINKED ISSUE #4414: Sample with Consistency Model
### Describe the bug

When I use diffusers to sample images with offered code by Consistency Model, I can't get an image that corresponds to the label at all: 
label 207 (Golden Retriever)

### Reproduction

I just used the officially offered code by Consistency Model:
https://github.com/openai/consistency_models

### Logs

```shell
python /home/dqwang/scratch/tutu/diffuser/check.py
unet/diffusion_pytorch_model.safetensors not found
Loading pipeline components...:   0%|                                                     | 0/2 [00:00<?, ?it/s]Some weights of the model checkpoint were not used when initializing UNet2DModel: 
 ['mid_block.attentions.0.group_norm.weight, mid_block.attentions.0.group_norm.bias']
Loading pipeline components...: 100%|█████████████████████████████████████████████| 2/2 [00:02<00:00,  1.11s/it]
/home/dqwang/mambaforge/envs/diffuser/lib/python3.7/site-packages/diffusers/configuration_utils.py:134: FutureWarning: Accessing config attribute `sigma_min` directly via 'CMStochasticIterativeScheduler' object attribute is deprecated. Please access 'sigma_min' over 'CMStochasticIterativeScheduler's config object instead, e.g. 'scheduler.config.sigma_min'.
  deprecate("direct config name access", "1.0.0", deprecation_message, standard_warn=False)
100%|█████████████████████████████████████████████████████████████████████████████| 3/3 [00:00<00:00, 14.55it/s]
```

### System Info

- `diffusers` version: 0.19.3
- Platform: Linux-5.15.0-25-generic-x86_64-with-debian-bookworm-sid
- Python version: 3.7.12
- PyTorch version (GPU?): 1.11.0+cu113 (True)
- Huggingface_hub version: 0.16.4
- Transformers version: not installed
- Accelerate version: 0.20.3
- xFormers version: not installed
- Using GPU in script?: Yes
- Using distributed or parallel set-up in script?: No

### Who can help?

_No response_

### PR DESCRIPTION
**Pull request by dg845**
# What does this PR do?

This PR aims to fix a consistency models bug in the `UNet2DMidBlock2D` block, where we want to set both `resnet_time_scale_shift == scale_shift` and the internal `Attention` block to have a `torch.nn.GroupNorm` layer.

The current behavior is to only create the `GroupNorm` layer if `resnet_time_scale_shift == default`, but some consistency model checkpoints (in particular, the ImageNet-64 checkpoints) have `resnet_time_scale_shift != default` and a group norm in their mid block attention layer.

Currently, the PR fixes this issue by adding a `attn_groups` argument to `UNet2DMidBlock2D`, which allows the user to directly configure the `GroupNorm` layer in `UNet2DMidBlock2D`'s `Attention` block (and exposing it in `UNet2DModel`). If `attn_groups` is set to an integer, the `GroupNorm` will (always) be created with the given number of groups. If it is set to `None`, the current behavior is retained. This should allow previous `UNet2DModel` checkpoints to work as expected without having to modify core code.

Fixes #4414. Based on a previous PR #4437.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@patrickvonplaten 
@sayakpaul 
@RossoneriZhao

### CODE REVIEW

File: src/diffusers/models/unet_2d.py [hunk 1]
```diff
@@ -105,6 +105,7 @@ def __init__(
         act_fn: str = "silu",
         attention_head_dim: Optional[int] = 8,
         norm_num_groups: int = 32,
+        attn_norm_num_groups: Optional[int] = None,
```
Comment (sayakpaul) [inline src/diffusers/models/unet_2d.py:108]: Add a docstring entry?

**Comment by sayakpaul** [comment]
Thanks for the PR! 

Is it also possible to do some testing to ensure correctness?

File: src/diffusers/models/unet_2d.py (hunk 1 above)
Comment (dg845) [reply src/diffusers/models/unet_2d.py:112]: I have added a docstring entry for `attn_norm_num_groups`. It looks like none of the `UNetMidBlock2D` arguments have docstrings, so I didn't add one for `attn_groups`.

**Comment by dg845** [comment]
I have tested the new consistency model checkpoints locally, and the results look reasonable:

`cd_imagenet64_lpips` onestep generation:

cm_golden_retriever_onestep

`cd_imagenet64_lpips` onestep generation:

cm_golden_retriever_multistep

Unfortunately for some reason when I try to upload the new checkpoints to the hub via the website it looks like the hashing step hangs indefinitely:

cm_ckpt_hashing

The progress bar has been stuck like that for a long time without making progress. (I think I could have only replaced `unet/config.json`, but I already deleted the `unet` so I need to re-upload it.)

**Comment by sayakpaul** [comment]
Could you try using the `push_to_hub()` method directly on the model? 

Docs: https://huggingface.co/docs/diffusers/using-diffusers/push_to_hub#:~:text=%F0%9F%A4%97%20Diffusers%20provides%20a%20PushToHubMixin,share%20your%20work%20with%20others.. 

@DN6 does it make sense to add a slow test here to ensure the correctness of this feature?

**Comment by dg845** [comment]
Using `push_to_hub()` worked, thanks!

The following consistency models checkpoints have been updated:

- [`dg845/consistency-model-pipelines`](https://huggingface.co/dg845/consistency-model-pipelines/tree/main) (corresponds to `cd_imagenet64_l2`)
- [`dg845/diffusers-cd_imagenet64_lpips`](https://huggingface.co/dg845/diffusers-cd_imagenet64_lpips/tree/main)
- [`dg845/diffusers-ct_imagenet64`](https://huggingface.co/dg845/diffusers-ct_imagenet64/tree/main)

If the changes look good, we will have to propagate the changes to the corresponding `openai/*` repos, as well as to [`diffusers/consistency_models`](https://huggingface.co/diffusers/consistency_models/tree/main) for the slow tests.

**Comment by dg845** [comment]
I have added a test for the `attn_norm_num_groups` argument to `UNet2DModelTests`.

**Comment by sayakpaul** [comment]
Changes look good to me, thanks a lot! 

@patrickvonplaten could you take care of https://github.com/huggingface/diffusers/pull/4863#issuecomment-1702310161?

**Comment by dg845** [comment]
For reasons I don't yet understand, the expected slices for the consistency models seem to have changed. Rerunning the scripts to get the expected slices (e.g. `cm/image_sample.py` on [`dg845/consistency_models:local_test_cuda`](https://github.com/dg845/consistency_models/tree/local_test_cuda)) and using those expected slices was necessary to make the slow tests pass again. (I'm currently using [`dg845/consistency-model-pipelines`](https://huggingface.co/dg845/consistency-model-pipelines/tree/main) for local testing.)

**Comment by dg845** [comment]
Gentle ping @patrickvonplaten: could you help upload the new checkpoints to the below repositories? The full map of changes is as follows (I have submitted PRs where possible):

- [`dg845/consistency-models-test`](https://huggingface.co/dg845/consistency-models-test) -> [`diffusers/consistency-models-test`](https://huggingface.co/diffusers/consistency-models-test/tree/main) [[PR](https://huggingface.co/diffusers/consistency-models-test/discussions/1)]
- [`dg845/consistency-model-pipelines/unet`](https://huggingface.co/dg845/consistency-model-pipelines/tree/main/unet) -> [`diffusers/consistency_models/diffusers_cd_imagenet64_l2`](https://huggingface.co/diffusers/consistency_models/tree/main/diffusers_cd_imagenet64_l2) [*]
- [`dg845/consistency-model-pipelines`](https://huggingface.co/dg845/consistency-model-pipelines/tree/main) -> [`openai/diffusers-cd_imagenet64_l2`](https://huggingface.co/openai/diffusers-cd_imagenet64_l2) [[PR](https://huggingface.co/openai/diffusers-cd_imagenet64_l2/discussions/1)]
- [`dg845/diffusers-cd_imagenet64_lpips`](https://huggingface.co/dg845/diffusers-cd_imagenet64_lpips/tree/main) -> [`openai/diffusers-cd_imagenet64_lpips`](https://huggingface.co/openai/diffusers-cd_imagenet64_lpips) [[PR](https://huggingface.co/openai/diffusers-cd_imagenet64_lpips/discussions/1)]
- [`dg845/diffusers-ct_imagenet64`](https://huggingface.co/dg845/diffusers-ct_imagenet64/tree/main) -> [`openai/diffusers-ct_imagenet64`](https://huggingface.co/openai/diffusers-ct_imagenet64) [[PR](https://huggingface.co/openai/diffusers-ct_imagenet64/discussions/1)]

[*] I haven't submitted a PR for this change since it involves changing a subfolder of the directory, which isn't supported by [`push_to_hub()`](https://huggingface.co/docs/diffusers/v0.20.0/en/api/pipelines/overview#diffusers.utils.PushToHubMixin.push_to_hub). Alternatively, we could change [`diffusers/consistency_models`](https://huggingface.co/diffusers/consistency_models) to hold the whole pipeline checkpoint instead of just the U-Net, or directly use [`openai/diffusers-cd_imagenet64_l2`](https://huggingface.co/openai/diffusers-cd_imagenet64_l2) in the slow tests, which I believe is the same underlying checkpoint.

**Review by patrickvonplaten — APPROVED** [review]
Changes lgtm - @sayakpaul wdyt?

**Review by sayakpaul — APPROVED** [review]
LGTM given the CI is green. Thank you for your great work!

Could you help me understand why are we changing the assertion values? You must have explained it somewhere but I am unable to recollect it (apologies).

**Comment by dg845** [comment]
> Could you help me understand why are we changing the assertion values?

When I reran the slow tests on the `cd_imagenet64_l2` checkpoint with the new `UNet2DModel` config, the tests failed. Rerunning the scripts to get new `expected_slice`s without changing either the original consistency models code nor any consistency model code in `diffusers` (besides the PR change) and using those in the failing tests caused them to pass.

I'm fairly (although not 100%) certain that the reason is error on my part when I added the original `expected_slice`s: somehow I used expected slices which were close to the output of the bugged models, but not actually correct. It seems like the PR change pushed the model output towards the true `expected_slice`s and out of range of the incorrect `expected_slice`s, which caused the tests to fail.

**Comment by sayakpaul** [comment]
Works for me! Thanks for explaining. 

Failing test is unrelated.
